tempesta_fw/http2/hpack_decoder.c

```diff
@@ -0,0 +1,648 @@
+/**
+ *		Tempesta FW
+ *
+ * HPACK (RFC-7541) decoder.
+ *
+ * Copyright (C) 2017 Tempesta Technologies, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+
+#include <stdint.h>
+#include <string.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include "common.h"
+#include "../pool.h"
+#include "../str.h"
+#include "../http.h"
+#include "errors.h"
+#include "buffers.h"
+#include "huffman.h"
+#include "hindex.h"
+#include "hpack_helpers.h"
+#include "hpack.h"
+
+#define Debug_HPack 1
+
+#if Debug_HPack
+#define DPRINTF(...) printf("HPack: " __VA_ARGS__)
+#define DPUTS(...) puts("HPack: " __VA_ARGS__)
+#else
+#define DPRINTF(...)
+#define DPUTS(...)
+#endif
+
+#define HPACK_COPY_STRING(value)    \
+do {				    \
+	n -= length;		    \
+	src = buffer_copy(	    \
+		source, &m, src, m, \
+		length, buffer, rc  \
+	);			    \
+	if (unlikely(* rc)) {	    \
+		goto Bug;	    \
+	}			    \
+	field->value = buffer->str; \
+} while (0)
+
+HTTP2Field *
+hpack_decode(HPack * __restrict hp,
+	     HTTP2Input * __restrict source,
```

Comment:
It seems that 'source' must me passed here with already initialized (i.e. parsed) `TfwStr` (as underlying instance for `HTTP2Input`), but considering HTTP/2 logic - we must initially decode HTTP/2 message and only after that pass it to HTTP parser, where `TfwStr` instances will be initialized.